File: 12_2.py
import math

directions = {"N": 0 + 1j, "S": 0 - 1j, "E": 1 + 0j, "W": -1 + 0j}


class Waypoint:
    def __init__(self, pos):
        self.pos = pos

    def move(self, dir, value):
        self.pos += directions[dir] * value

    def rotate(self, angle):
        x = math.cos(angle) * self.pos.real - math.sin(angle) * self.pos.imag
        y = math.sin(angle) * self.pos.real + math.cos(angle) * self.pos.imag
        self.pos = complex(x, y)


class Ship:
    def __init__(self, pos):
        self.pos = pos

    def move(self, waypoint, times):
        for i in range(times):
            self.pos += waypoint.pos


if __name__ == "__main__":
    with open("12.in") as f:
        input = [x.strip() for x in f.readlines()]

    s = Ship(0 + 0j)
    w = Waypoint(10 + 1j)
    for x in input:
        action = x[0]
        value = int(x[1:])
        if action in "NSEW":
            w.move(action, value)
        elif action in "LR":
            if action == "R":
                value *= -1
            value *= math.pi / 180
            w.rotate(value)
        elif action == "F":
            s.move(w, value)

    print(s.pos)
    print(abs(s.pos.real) + abs(s.pos.imag))

File: 12-2.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"os"
	"strconv"
	"strings"
)

func rotate(v complex128, degrees float64) complex128 {
	r := degrees * math.Pi / 180
	x := math.Cos(r)*real(v) - math.Sin(r)*imag(v)
	y := math.Sin(r)*real(v) + math.Cos(r)*imag(v)
	return complex(x, y)
}

func main() {
	bytes, err := ioutil.ReadFile("12.in")
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	lines := strings.Split(strings.TrimSpace(string(bytes)), "\n")

	ship := 0 + 0i
	waypoint := 10 + 1i
	for _, x := range lines {
		action := x[0]
		value, _ := strconv.ParseFloat(x[1:], 64)
		v := complex(value, 0)
		switch action {
		case 'N':
			waypoint += v * (0 + 1i)
		case 'S':
			waypoint += v * (0 - 1i)
		case 'E':
			waypoint += v * (1 + 0i)
		case 'W':
			waypoint += v * (-1 + 0i)
		case 'F':
			ship += v * waypoint
		case 'L':
			waypoint = rotate(waypoint, value)
		case 'R':
			waypoint = rotate(waypoint, -value)
		}
	}

	fmt.Println(math.Abs(real(ship)) + math.Abs(imag(ship)))
}
